File: app/src/main/java/com/odnovolov/randomuser/utils/BackPressActivity.java
package com.odnovolov.randomuser.utils;

import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;

public class BackPressActivity extends AppCompatActivity {
    public interface OnBackPressInterceptor {
        boolean onBackPressed();
    }

    private final ArrayList<OnBackPressInterceptor> backPressInterceptors = new ArrayList<>();

    public void registerOnBackPressInterceptor(OnBackPressInterceptor onBackPressInterceptor) {
        backPressInterceptors.add(onBackPressInterceptor);
    }

    public void unregisterOnBackPressInterceptor(OnBackPressInterceptor onBackPressInterceptor) {
        backPressInterceptors.remove(onBackPressInterceptor);
    }

    @Override
    public void onBackPressed() {
        for (OnBackPressInterceptor onBackPressInterceptor : backPressInterceptors) {
            boolean intercepted = onBackPressInterceptor.onBackPressed();
            if (intercepted) return;
        }
        super.onBackPressed();
    }
}

File: app/src/main/java/com/odnovolov/randomuser/BackPressActivity.kt
package com.odnovolov.randomuser

import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner

open class BackPressActivity : AppCompatActivity() {
    private val backPressInterceptors: MutableList<() -> Boolean> = ArrayList()

    fun registerOnBackPressed(onBackPressed: () -> Boolean) {
        backPressInterceptors.add(onBackPressed)
    }

    fun unregisterOnBackPressed(onBackPressed: () -> Boolean) {
        backPressInterceptors.remove(onBackPressed)
    }

    override fun onBackPressed() {
        backPressInterceptors.forEach { onBackPressed ->
            val intercepted = onBackPressed()
            if (intercepted) return
        }
        super.onBackPressed()
    }
}

fun Fragment.setOnBackPressed(action: () -> Boolean) {
    lifecycle.addObserver(object : DefaultLifecycleObserver {
        override fun onResume(owner: LifecycleOwner) {
            super.onResume(owner)
            (requireActivity() as BackPressActivity).registerOnBackPressed(action)
        }

        override fun onPause(owner: LifecycleOwner) {
            super.onPause(owner)
            (requireActivity() as BackPressActivity).unregisterOnBackPressed(action)
        }
    })
}